File: SpiralMatrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        
        vector<int > ans;
        
        int t = 0, b = matrix.size()-1, l = 0, r = matrix[0].size()-1, d = 0;
        
        while(t <= b && l <= r){
            
            //vector<int> row;
            if(d == 0){
                
                for(int i = l; i <= r;i++){
                    ans.push_back(matrix[t][i]);
                }
                t++;
            }
            else if(d == 1){
                
                for(int i = t; i <= b;i++){
                    ans.push_back(matrix[i][r]);
                }
                r--;
            }
            else if(d == 2){
                for(int i = r; i >= l;i--){
                    ans.push_back(matrix[b][i]);
                }
                b--;
            }
            else if(d == 3){
                for(int i = b; i >= t;i--){
                    ans.push_back(matrix[i][l]);
                }
                l++;
            }
            
            //ans.push_back(row);
            
            d = (d+1)%4;
        }
        
        return ans;
        
    }
};

File: SpiralMatrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        
        List<Integer> ans = new ArrayList<Integer>();

        int m = matrix.length;
        int n = matrix[0].length;
        int top = 0, bottom = m - 1, left = 0, right = n - 1, d = 0;

        while(top <= bottom && left <= right){
            
            switch(d){
                case 0:
                    for(int i = left;i <= right;i++)
                        ans.add(matrix[top][i]);
                    top++;
                    break;
                case 1:
                    for(int i = top;i <= bottom;i++)
                        ans.add(matrix[i][right]);
                    right--;
                    break;
                case 2:
                    for(int i = right;i >= left;i--)
                        ans.add(matrix[bottom][i]);
                    bottom--;
                    break;
                case 3:
                    for(int i = bottom; i >= top;i--)
                        ans.add(matrix[i][left]);
                    left++;
                    break;
            }
            d = (d+1)%4;
        }

        return ans;
    }
}
